File: engine/beatmarks_engine/envelope.py
"""音量エンベロープ(全体+3帯域、100Hz)と静寂検出。

- RMS を hop 256 で計算し dB 化 → −60〜0dB を 0〜1 に正規化 → 100Hz に補間
- 静寂検出は dB 系列へのしきい値+最小継続時間判定。UI 側も同じアルゴリズムを
  TS で持ち、しきい値変更は再解析なしで済む(スペック §5)
"""
import librosa
import numpy as np

from .hits import BANDS, band_filter

ENV_RATE = 100.0
SILENCE_DB = -45.0
SILENCE_MIN_DUR = 0.7

_RMS_HOP = 256
_RMS_FRAME = 1024
_DB_FLOOR = -120.0


def _rms_series(yb: np.ndarray, sr: int, grid: np.ndarray):
    rms = librosa.feature.rms(y=yb, frame_length=_RMS_FRAME, hop_length=_RMS_HOP)[0]
    times = librosa.frames_to_time(np.arange(len(rms)), sr=sr, hop_length=_RMS_HOP)
    db = 20.0 * np.log10(np.maximum(rms, 10 ** (_DB_FLOOR / 20.0)))
    norm = np.clip((db + 60.0) / 60.0, 0.0, 1.0)
    return np.interp(grid, times, norm), np.interp(grid, times, db)


def compute_envelopes(y: np.ndarray, sr: int) -> tuple[dict, np.ndarray]:
    dur = len(y) / sr
    grid = np.arange(0.0, dur, 1.0 / ENV_RATE)
    total_norm, total_db = _rms_series(y, sr, grid)
    env: dict = {"sampleRateHz": 100, "total": [float(v) for v in total_norm]}
    for band, (lo, hi) in BANDS.items():
        norm, _db = _rms_series(band_filter(y, sr, lo, hi), sr, grid)
        env[band] = [float(v) for v in norm]
    return env, total_db


def detect_silences(total_db: np.ndarray, threshold_db: float = SILENCE_DB,
                    min_dur_sec: float = SILENCE_MIN_DUR,
                    rate_hz: float = ENV_RATE) -> list[dict]:
    below = np.asarray(total_db) < threshold_db
    regions: list[dict] = []

    def _push(a: int, b: int) -> None:
        if (b - a) / rate_hz >= min_dur_sec:
            regions.append({
                "startSec": a / rate_hz,
                "endSec": b / rate_hz,
                "floorDb": float(np.min(total_db[a:b])),
            })

    start: int | None = None
    for i, flag in enumerate(below):
        if flag and start is None:
            start = i
        elif not flag and start is not None:
            _push(start, i)
            start = None
    if start is not None:
        _push(start, len(below))
    return regions

File: app/src/shared/envelope.ts
/** エンベロープ(100Hz正規化系列)のフレーム再サンプルと、静寂の再フィルタ。
 *  エンジンは既定しきい値の silences と生の envelope を返す。しきい値を UI で
 *  変えたときはここで envelope から再導出する(再解析不要。スペック §5)。 */
import { frameToTime, timeToFrame } from "./timebase.js";
import type { Fps, SilenceInfo } from "./types.js";

export function normToDb(norm: number): number {
  return norm * 60 - 60;
}

export function detectSilencesFromEnvelope(
  total: number[], rateHz: number, thresholdDb: number, minDurSec: number,
): SilenceInfo[] {
  const regions: SilenceInfo[] = [];
  let start: number | null = null;
  let minDb = Infinity;

  const push = (a: number, b: number, floor: number) => {
    if ((b - a) / rateHz >= minDurSec) {
      regions.push({ startSec: a / rateHz, endSec: b / rateHz, floorDb: floor });
    }
  };

  for (let i = 0; i < total.length; i++) {
    const db = normToDb(total[i]!);
    if (db < thresholdDb) {
      if (start === null) {
        start = i;
        minDb = db;
      } else if (db < minDb) {
        minDb = db;
      }
    } else if (start !== null) {
      push(start, i, minDb);
      start = null;
      minDb = Infinity;
    }
  }
  if (start !== null) push(start, total.length, minDb);
  return regions;
}

/** フレーム0..N(N = floor(dur×fps))ごとにエンベロープ値を線形補間で返す。 */
export function resampleEnvelopeToFps(
  env: number[], rateHz: number, fps: Fps, durationSec: number,
): number[] {
  const lastFrame = timeToFrame(durationSec, fps, "floor");
  const out: number[] = new Array(lastFrame + 1);
  for (let f = 0; f <= lastFrame; f++) {
    const t = frameToTime(f, fps);
    const x = t * rateHz;
    const i = Math.floor(x);
    const frac = x - i;
    const a = env[Math.min(i, env.length - 1)] ?? 0;
    const b = env[Math.min(i + 1, env.length - 1)] ?? a;
    out[f] = a + (b - a) * frac;
  }
  return out;
}
